Revisited endpoint state restore

aanm opened the pull request:
Fixes #1140

aanm committed:
daemon: spawning sync docker after sleep

Since SyncDocker was triggered before calling
RunBackgroundContainerSync, it doesn't make sense to call SyncDocker
right after executing RunBackgroundContainerSync.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
daemon: revisited restore functionality

Bug fixing that where preventing endpoints from being restored.

This functionality needs to be fully tested, following commits
will provide unit and runtime tests.

Signed-off-by: André Martins <andre@cilium.io>

ianvernon commented on daemon/state.go:
```diff
@@ -219,8 +226,8 @@ func (d *Daemon) syncLabels(ep *endpoint.Endpoint) error {
 
-// cleanUpDockerDandlingEndpoints cleans all endpoints that are dandling by checking out
-// if a particular endpoint has its container running.
+// cleanUpDockerDandlingEndpoints cleans all endpoints that are dandling by
```
Nit: Dandling --> Dangling

aanm replied:
Fixed!

aanm committed:
daemon: regenerate endpoints in StateDir

changed the base path of the endpoints to be the StateDir. This will
allow to specify any directory to unit test particular functions that
use the owner interface.

Signed-off-by: André Martins <andre@cilium.io>

tgraf commented on daemon/state.go:
```diff
@@ -47,67 +48,73 @@ func (d *Daemon) SyncState(dir string, clean bool) error {
+		ep.Mutex.Unlock()
 	}
 
 	log.Infof("Restored %d endpoints", restored)
```
This number of restored endpoints seems premature as it may fail later. I suggest we remove it.

tgraf commented on daemon/state.go:
```diff
@@ -47,67 +48,73 @@ func (d *Daemon) SyncState(dir string, clean bool) error {
-		}
+		go func(ep *endpoint.Endpoint, wg *sync.WaitGroup) {
+			if err := d.allocateIPs(ep); err != nil {
+				log.Errorf("Failed while reallocating ep %d's IP addresses: %s. Endpoint won't be restored", ep.ID, err)
```
Let's be consistent and call it endpoint in all info messages
`"Failed to re-allocate IP of endpoint %d. Not restoring endpoint."`

tgraf commented on daemon/state.go:
```diff
@@ -47,67 +48,73 @@ func (d *Daemon) SyncState(dir string, clean bool) error {
-			ep.Mutex.RUnlock()
-		}
-		log.Infof("Restored endpoint %d", epID)
+			log.Infof("EP %d's IP addresses successfully reallocated", ep.ID)
```
This info message seems useless as it again may fail afterwards.

tgraf commented on daemon/state.go:
```diff
@@ -47,67 +48,73 @@ func (d *Daemon) SyncState(dir string, clean bool) error {
+			} else {
+				ep.Mutex.RUnlock()
+			}
+			log.Infof("Restored endpoint %d", epID)
```
This is the most important info message, we may want to report the IP addresses as well.

tgraf commented on daemon/state.go:
```diff
@@ -47,67 +48,73 @@ func (d *Daemon) SyncState(dir string, clean bool) error {
 	endpointmanager.Mutex.Unlock()
 
+	wg.Wait()
+
```
`"Restored %d endpoints"`

aanm replied:
Change it a little bit, better?

```
"Restored endpoint 38939 with IPs 10.15.250.189 and f00d::a00:20f:0:981b"
"Found 3 endpoints on which 3 were restored"
```

tgraf replied:
nit: `.. of which 3 were restored`

tgraf replied:
Otherwise looks great now!

aanm replied:
Fixed

tgraf commented on daemon/docker_watcher.go:
```diff
@@ -89,8 +89,8 @@ func (d *Daemon) SyncDocker() {
 	for {
-		d.SyncDocker()
 		time.Sleep(syncRateDocker)
+		d.SyncDocker()
```
Typo in commit title. "Spawning"

aanm committed:
daemon/state: removed ep.DockerID when re-syncing labels

Since RefCount was changed to the endpoint ID plus the host IP, the
ep.DockerID is not needed when restoring state from previous endpoints.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
daemon: add tests for restore state

Signed-off-by: André Martins <andre@cilium.io>

tgraf reviewed (changes requested):
Only a small nit left

aanm committed:
daemon: changing restore flag default value to true

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
tests: add restore state runtime tests in a k8s

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
Jenkinsfile: bumping timeout to 90 minutes

Signed-off-by: André Martins <andre@cilium.io>
